crates/accelerate/src/high_level_synthesis.rs

```diff
@@ -0,0 +1,252 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use hashbrown::HashMap;
+use pyo3::prelude::*;
+
+/// Track global qubits by their state.
+/// The global qubits are numbered by consecutive integers starting at `0`,
+/// and the states are distinguished into clean (:math:`|0\rangle`)
+/// and dirty (unknown).
+#[pyclass]
+#[derive(Clone, Debug)]
+pub struct QubitTracker {
+    /// The total number of global qubits
+    num_qubits: usize,
+    /// Stores the state for each qubit: `true` means clean, `false` means dirty
+    state: Vec<bool>,
+    /// Stores whether qubits are allowed be used
+    enabled: Vec<bool>,
+    /// Used internally for keeping the computations in `O(n)`
+    ignored: Vec<bool>,
+}
+
+#[pymethods]
+impl QubitTracker {
+    #[new]
+    pub fn new(num_qubits: usize) -> Self {
+        QubitTracker {
+            num_qubits,
+            state: vec![false; num_qubits],
+            enabled: vec![true; num_qubits],
+            ignored: vec![false; num_qubits],
+        }
+    }
+
+    /// Sets state of the given qubits to dirty
+    fn set_dirty(&mut self, qubits: Vec<usize>) {
+        for q in qubits {
+            self.state[q] = false;
+        }
+    }
+
+    /// Sets state of the given qubits to clean
+    fn set_clean(&mut self, qubits: Vec<usize>) {
+        for q in qubits {
+            self.state[q] = true;
+        }
+    }
+
+    /// Disables using the given qubits
+    fn disable(&mut self, qubits: Vec<usize>) {
+        for q in qubits {
+            self.enabled[q] = false;
+        }
+    }
+
+    /// Enable using the given qubits
+    fn enable(&mut self, qubits: Vec<usize>) {
+        for q in qubits {
+            self.enabled[q] = true;
+        }
+    }
+
+    /// Returns the number of enabled clean qubits, ignoring the given qubits
+    /// ToDo: check if it's faster to avoid using ignored
+    fn num_clean(&mut self, ignored_qubits: Vec<usize>) -> usize {
+        for q in &ignored_qubits {
+            self.ignored[*q] = true;
+        }
+
+        let count = (0..self.num_qubits)
+            .filter(|q| !self.ignored[*q] && self.enabled[*q] && self.state[*q])
+            .count();
+
+        for q in &ignored_qubits {
+            self.ignored[*q] = false;
+        }
+
+        count
+    }
```

Comment:
Just so you're able to test quickly, see if using a hashset affects this in any way:
```suggestion
    /// Returns the number of enabled clean qubits, ignoring the given qubits
    /// ToDo: check if it's faster to avoid using ignored
    fn num_clean(&mut self, ignored_qubits: HashSet<usize>) -> usize {
        (0..self.num_qubits)
            .filter(|q| !ignored_qubits.contains(q) && self.enabled[*q] && self.state[*q])
            .count()
    }
```